File: src/leetCodeProblems/Bucketing/FindDuplicates442.go
package main

/*
- LeetCode - https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/
- Time - O(n)
- Space - O(1)
*/

func findDuplicates(nums []int) []int {
	n := len(nums) + 1

	out := []int{}

	for _, v := range nums {
		base := v % n
		nums[base-1] += n
	}

	for i, v := range nums {
		if v/n == 2 {
			out = append(out, i+1)
		}
	}
	return out
}

// func main() {
// 	log.Println(findDuplicates([]int{4, 3, 2, 7, 8, 2, 3, 1}))
// 	log.Println(findDuplicates([]int{1, 1, 2}))
// 	log.Println(findDuplicates([]int{1}))
// 	log.Println(findDuplicates([]int{10, 2, 5, 10, 9, 1, 1, 4, 3, 7}))
// }

File: src/leetCodeProblems/Bucketing-Inplace/FindDuplicates442.java
package leetCodeProblems.Bucketing;

/**
 * LeetCode - https://leetcode.com/problems/find-all-duplicates-in-an-array/
 * TimeComplexity - O(n)
 * SpaceComplexity - O(1)
 */

import java.util.*;

public class FindDuplicates442 {

	public List<Integer> findDuplicates(int[] nums) {

        List<Integer> output = new ArrayList<Integer>();

        for(int i=0; i<nums.length; i++) {

            // This is important, to get original value at nums[i].
            // Because it might be changed due to array operations
            int hashMapIndexInArray = nums[i] % nums.length;

            if (hashMapIndexInArray >= 0) {
                //System.out.println (hashMapIndexInArray);
                nums[hashMapIndexInArray] = nums[hashMapIndexInArray] + nums.length;
            }
        }

        for(int i=0; i<nums.length; i++) {

            //System.out.println(nums[i]);
            if (nums[i] > (nums.length*2)) {

                if (i == 0) {
                    output.add(nums.length);
                }
                else {
                    output.add(i);
                }

            }
        }

        return output;
    }

}
